### PR TITLE
Version and simple manifest

### PR DESCRIPTION
**Pull request by mrb**
/c @brynary

### CODE REVIEW

File: SPEC.md [hunk 1]
```diff
@@ -182,6 +183,33 @@ Contents give more information about the issue's check, including a description
...
+}
+```
+
+The fields in the manifest include:
+
+* `name` (`String`) - the name of the package.
+* `maintainer` (`Object`)
+  * `name` (`String`) - the name of the maintainer
+  * `email` (`String`) - the email address of the maintainer
+* `spec_version` (`String`) - the most recent version of the specification which this engine supports
```
Comment (pbrisbin) [inline SPEC.md]: I find this a little confusing. What am I specifying here?

- I work with exactly and only 0.0.1
- I work with versions up to 0.0.1
- I work with versions at least 0.0.1
- Something else?

I think we need to support a limited concept of version bounds here. I don't want to sign up for crazy bounds checking logic, but we need to allow the authors to clearly show if it's newer-than, older-than, or exactly the `spec_version` they're specifying.

File: SPEC.md [hunk 2]
```diff
@@ -1,6 +1,7 @@
-# Code Climate Engine specification
+# Code Climate Engine Specification
+## Version 0.1
```
Comment (pbrisbin) [inline SPEC.md]: Seems risky / annoying to have this in two places. I vote it only be specified in the `VERSION` file.

File: SPEC.md (hunk 1 above)
Comment (mrb) [inline SPEC.md]: @pbrisbin Cool, which one do you think we should do? I don't have strong opinions here.

Comment (pbrisbin) [inline SPEC.md]: If we're intending to follow semantic versioning for the spec (which I'm in favor of), then I might suggest the following:

Keep the format as what you have, a single field with a single version value.

Interpret that value like the `~>` operator in gemspecs.

So:

```
spec_version: 0.0.1 means
  I (claim to) support >= 0.0.1 && < 0.1.0

spec_version: 0.1 means
  I support >= 0.1.0 && < 1.0.0

spec_version: 1 means
  I support everything (>= 1.0.0 && < *)
```

(I think I have that right, correct me if I'm wrong)

WDYT?

**Comment by wfleming** [comment]
@mrb While you're doing this, can you also please add a `version` or `engine_version` field as an integer? We're going to need it for cache implementation, so we may as well add it now.

File: SPEC.md [hunk 3]
```diff
@@ -182,6 +183,33 @@ Contents give more information about the issue's check, including a description
 
 Some engines require the ability to refer to other source locations. For this reason, the Issue type has an optional `other_locations` field, which is an array of other `Location` items that this issue needs to refer to.
 
+## Versioning
+
+This specification is versioned. The current version is [in the repository](https://github.com/codeclimate/spec/blob/master/VERSION). Engines declare the version of the specification they are compatible with in the manifest file, described below.
+
+## Manifest File
+
+All engines must include a `manifest.json` file at `/manifest.json`. This file, known as the manifest, includes information that is necessary for the analysis runtime. Here is an example of a manifest:
```
Comment (brynary) [inline SPEC.md]: I'd call it `engine.json` not `manifest.json`. A manifest is a list of contents, which this includes, but it also includes other things. (In other words, this file _contains_ a manifest, but is not a manifest itself)

File: SPEC.md [hunk 4]
```diff
@@ -182,6 +183,33 @@ Contents give more information about the issue's check, including a description
...
+    "email": "mrb@codeclimate.com",
+  },
+  "spec_version": "0.0.1",
+}
+```
+
+The fields in the manifest include:
+
+* `name` (`String`) - the name of the package.
+* `maintainer` (`Object`)
```
Comment (brynary) [inline SPEC.md]: To @wfleming's point, `version` and `description` should be required fields also.

File: SPEC.md (hunk 1 above)
Comment (brynary) [inline SPEC.md]: I actually think this should be a fixed version. It is the version that you are certifying your engine conforms to. If we have another, newer, spec version, it is incumbent on us to determine whether it is backwards compatible.

Comment (pbrisbin) [inline SPEC.md]: I don't fully understand what you're saying, but if you can picture a future where engines specify an exact spec version they conform to and we don't end up in dependency hell as the spec evolves and cli/builder start checking and relying on said conformance, then :+1: from me.

Comment (pbrisbin) [inline SPEC.md]: We can actually play this out using the `include_path` feature as an example:

- Pretend it's a few weeks ago, before we built the `include_paths` feature
- Pretend at that time the spec was v0.1
- Pretend we had this engines.json thing and all engines have `spec_version: 0.1`
- We implement `include_paths` and deprecate `exclude_paths` in SPEC
- What happens?
- We actually want to stop supporting `exclude_paths` in CLI/Builder
- What happens?

Comment (brynary) [inline SPEC.md]: The answer to the first "What happens?" is:

* We bump the spec version to 1.0 (not backwards compatible). The Code Climate CLI, during that time, supports both the 1.x and 0.x SPEC. When it runs an engine, it looks at the declared `spec_version`, and if it's `0.x` it runs it with the old engine harness (probably emitting a warning)

The answer to the second "What happens?" is that we remove the `0.x` engine harness from the CLI, and it changes to refuse to run any pre-`1.x`-designated engines

Comment (pbrisbin) [inline SPEC.md]: Cool cool. This is the first I'm hearing of this versioned Engine Harness idea, and it sounds pretty non-trivial -- but OK, that all sounds fine on paper assuming we build said harness.

File: SPEC.md [hunk 5]
```diff
@@ -182,6 +182,37 @@ Contents give more information about the issue's check, including a description
...
+
+The fields in the manifest include:
+
+* `name` (`String`) - the name of the package
+* `description` (`String`) - a description of the engine
+* `maintainer` (`Object`) - data about the engine maintainer
+  * `name` (`String`) - the name of the maintainer
+  * `email` (`String`) - the email address of the maintainer
+* `version` (`Integer`) - engine version
+* `spec_version` (`String`) - the most recent version of the specification which this engine supports
```
Comment (pbrisbin) [inline SPEC.md]: Still confused by "the most recent version...", why not just "the version..."?

File: SPEC.md [hunk 6]
```diff
@@ -182,6 +182,37 @@ Contents give more information about the issue's check, including a description
...
+```
+
+The fields in the manifest include:
+
+* `name` (`String`) - the name of the package
+* `description` (`String`) - a description of the engine
+* `maintainer` (`Object`) - data about the engine maintainer
+  * `name` (`String`) - the name of the maintainer
+  * `email` (`String`) - the email address of the maintainer
+* `version` (`Integer`) - engine version
```
Comment (pbrisbin) [inline SPEC.md]: Nit: by why are you forcing engines into an integer versioning scheme?

File: SPEC.md (hunk 5 above)
Comment (mrb) [inline SPEC.md]: Good catch

File: SPEC.md (hunk 6 above)
Comment (mrb) [inline SPEC.md]: @pbrisbin At @wfleming's request. What do you prefer?

Comment (pbrisbin) [inline SPEC.md]: Just a string, like any version.

Comment (wfleming) [inline SPEC.md]: Bryan had a strong preference for integer versioning. I'm not too fussed. For cache invalidation (which is what I care about), either works.

Comment (brynary) [inline SPEC.md]: Ah, sorry. In thinking about it, I think so long as we roll the cache on any change to the version, a string is fine, and engines can pick their own format.



—Bryan Helmkamp, Founder, Code Climate

bryan@brynary.com / 646-379-1810 / @brynary




Typed with thumbs.

File: SPEC.md [hunk 7]
```diff
@@ -182,6 +182,37 @@ Contents give more information about the issue's check, including a description
 
 Some engines require the ability to refer to other source locations. For this reason, the Issue type has an optional `other_locations` field, which is an array of other `Location` items that this issue needs to refer to.
 
+## Versioning
+
+This specification is versioned. The current version is [in the repository](https://github.com/codeclimate/spec/blob/master/VERSION). Engines declare the version of the specification they are compatible with in the manifest file, described below.
+
+## Engine Config File
+
+All engines must include a `engine.json` file at `/engine.json`. This file includes information that is necessary for the analysis runtime and metadata about the engine. Here is an example config:
```
Comment (wfleming) [inline SPEC.md]: grammar nit: "an `engine.json`", not "a `engine.json`"

**Comment by mrb** [comment]
Okay, ready for final review @brynary @pbrisbin @wfleming

**Comment by brynary** [comment]
Looks good. Ship it

**Pull request merged**